lib/tpm2_hash.c

```diff
@@ -134,3 +134,130 @@ tool_rc tpm2_hash_file(ESYS_CONTEXT *ectx, TPMI_ALG_HASH halg,
     return tpm2_hash_common(ectx, halg, hierarchy, input, NULL, 0, result,
         validation);
 }
+
+static tool_rc tpm2_sm2_compute_msg_digest(ESYS_CONTEXT *ectx, TPMI_ALG_HASH halg,
+        TPMI_RH_HIERARCHY hierarchy, FILE *infilep, BYTE *inbuffer,
+        UINT16 inbuffer_len, TPM2B_DIGEST *z_digest, TPM2B_DIGEST **result,
+        TPMT_TK_HASHCHECK **validation)
+{
+    bool use_left = true, done;
+    unsigned long left = inbuffer_len;
+    size_t bytes_read;
+    TPM2B_AUTH null_auth = TPM2B_EMPTY_INIT;
+    TPMI_DH_OBJECT sequence_handle;
+    TPM2B_MAX_BUFFER buffer;
+
+    /*  if we're using infilep, get file size */
+    if (!!infilep) {
+        /* Suppress error reporting with NULL path */
+        use_left = files_get_file_size(infilep, &left, NULL);
+    }
+
+    buffer.size = BUFFER_SIZE(typeof(buffer), buffer);
+    if (z_digest->size > buffer.size) {
+        return tool_rc_general_error;
+    }
+
+    buffer.size = z_digest->size;
+    memcpy(buffer.buffer, z_digest->buffer, buffer.size);
```

Comment:
Can you refactor the code to include this only difference with tpm2_hash_common and simply call tpm2_hash_common instead of duplicating most of the function. Likewise for tpm2_sm2_compute_msg_digest_data and tpm2_sm2_compute_msg_digest_file.

Reply:
sm2 needs hash(z || message),  tpm2_hash_common is not applicable if message size is so big.